cpp.mk

```diff
@@ -0,0 +1,181 @@
+
+endef
+
+# Extra rules for TEST_TARGETS
+#
+# @param $1 - Target name (final executable)
+define test_target
+
+# Call into standard build infrastructure for $1
+$(call common_target,$1,test)
+
+$${BUILD_DIR}/bin/$1 : $${BUILD_DIR}/obj/gmock-all.o $${BUILD_DIR}/obj/gtest-all.o
+
+.PHONY : run_$1
+run_$1 : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ $$< ${EXTRA_TEST_ARGS}
+
+# This sentinel file proves the tests have *all* been run on this build (mostly for coverage)
+$${BUILD_DIR}/$1/.$1_already_run : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ $$<
+	@touch $$@
+
+.PHONY : debug_$1
+debug_$1 : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ gdb --args $$< $${EXTRA_TEST_ARGS}
+
+$${BUILD_DIR}/$1/valgrind_output.xml : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ valgrind --xml=yes --xml-file=$$@ $$<
+
+.PHONY : valgrind_check_$1
+valgrind_check_$1 : $${BUILD_DIR}/$1/valgrind_output.xml
+	@mkdir -p $${BUILD_DIR}/scratch/
+	@xmllint --xpath '//error/kind' $$< 2>&1 | \
+		sed -e 's#<kind>##g' | \
+		sed -e 's#</kind>#\n#g' | \
+		sort > $${BUILD_DIR}/scratch/valgrind.tmp
+	@if grep -q -v "XPath set is empty" $${BUILD_DIR}/scratch/valgrind.tmp ; then \
+		echo "Error: some memory errors have been detected" ; \
+		cat $${BUILD_DIR}/scratch/valgrind.tmp ; \
+		echo "See $$< for further details." ; \
+		exit 2 ; \
+	fi
+
+.PHONY : valgrind_$1
+valgrind_$1 : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ valgrind $$< $${EXTRA_TEST_ARGS}
+
+# Coverage arguments for $1
+$1_COVERAGE_ARGS := --object-directory=$(shell pwd) --root $$(shell pwd) --exclude "^ut" $${$1_OBJECT_DIR}
+
+$${BUILD_DIR}/$1/coverage.xml : $${BUILD_DIR}/$1/.$1_already_run
+	@../modules/gcovr/scripts/gcovr $${$1_COVERAGE_ARGS} --xml > $$@
+
+.PHONY : coverage_check_$1
+coverage_check_$1 : $${BUILD_DIR}/$1/coverage.xml
+	@mkdir -p $${BUILD_DIR}/scratch/
+	@xmllint --xpath '//class[@line-rate!="1.0"]/@filename' $$< \
+		| tr ' ' '\n' \
+		| grep filename= \
+		| cut -d\" -f2 \
+		| sort > $${BUILD_DIR}/scratch/coverage_$1.tmp
+	@sort ut/coverage-not-yet | comm -23 $${BUILD_DIR}/scratch/coverage_$1.tmp - > $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp
+	@if grep -q ^ $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp ; then \
+		echo "Error: some files unexpectedly have less than 100% code coverage:" ; \
+		cat $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp ; \
+		exit 2 ; \
+	fi
+CLEANS += $${BUILD_DIR}/scratch/coverage_$1.tmp $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp
+
+.PHONY : coverage_raw_$1
+coverage_raw_$1 : $${BUILD_DIR}/$1/.$1_already_run
+	@../modules/gcovr/scripts/gcovr $${$1_COVERAGE_ARGS} --keep --sort-percentage
+
+test : run_$1
+valgrind : valgrind_$1
+vg_check : valgrind_check_$1
+coverage_check : coverage_check_$1
+coverage_raw : coverage_raw_$1
+
+CLEANS += $${BUILD_DIR}/$1/valgrind_output.xml $${BUILD_DIR}/$1/.$1_already_run
+
+endef
+
+# Default values for build flags for each build
+__COMMON_CPPFLAGS := -ggdb3 -std=c++11
```

Comment:
Also, nice try at sneaking it through, but we're `-std=c++0x` at the moment so this change at least warrants a team heads-up.